### PR TITLE
ID-1313 resource type admin policy inheritance

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: https://broadworkbench.atlassian.net/browse/ID-1313

[Design doc](https://docs.google.com/document/d/13Mhiuym1zh7lOQ4VOeEbenS1WqmHL5v1cn_fUZwpR9o/edit#heading=h.6mdcf0m901wb)

Also, deleting a resource was refactored to put all database operations in a single transaction and to do a foreign key violation check before anything destructive happens. Had to touch this code anyway to remove effective policies from deleted resources where we want to leave a tomb stone record.

What:

implement resource policies that are inherited from resource type admin of the associated type

Why:

we intend to use this to grant Terra services blanket access to certain actions for all resources of a type

How:

Populate effective policy table from policies on resource type admin. There are 2 paths:
1. when a resource type admin policy is created or updated, effective policies for all associated resources need to be updated with those changes
2. when a resource is created, effective policies must be add from its resource type admin

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've filled out the [Security Risk Assessment](https://sdarq.dsp-appsec.broadinstitute.org/jira-ticket-risk-assesment) (requires Broad Internal network access) and attached the result to the JIRA ticket

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/AccessPolicyDAO.scala [hunk 1]
```diff
@@ -18,7 +18,10 @@ trait AccessPolicyDAO {
 
   def createResource(resource: Resource, samRequestContext: SamRequestContext): IO[Resource]
 
-  def deleteResource(resource: FullyQualifiedResourceId, samRequestContext: SamRequestContext): IO[Unit]
+  /** Deletes a resource. If leaveTombStone is true, a record of the resource will be kept in the database to prevent the resource name from being reused. If
+    * leaveTombStone is false, the resource will be completely removed from the database.
+    */
+  def deleteResource(resource: FullyQualifiedResourceId, leaveTombStone: Boolean, samRequestContext: SamRequestContext): IO[Unit]
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/AccessPolicyDAO.scala:24]: I love this, but why now?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala [hunk 2]
```diff
@@ -114,6 +124,60 @@ trait EffectivePolicyMutationStatements {
...
+    */
+  private def insertEffectivePolicyRolesFromResourceTypeAdmin(newResourcePK: ResourcePK, adminTypePK: ResourceTypePK)(implicit session: DBSession): Int = {
+    val erCol = EffectivePolicyRoleTable.column
+    val pr = PolicyRoleTable.syntax("pr")
+    val fr = FlattenedRoleMaterializedView.syntax("fr")
+    val rr = ResourceRoleTable.syntax("rr")
+    val ep = EffectiveResourcePolicyTable.syntax("ep")
+    val sourcePolicy = PolicyTable.syntax("source_policy")
+    val newResource = ResourceTable.syntax("newResource")
+    val adminResource = ResourceTable.syntax("adminResource")
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala]: Nit: Can we decide on acronyms, `snake_case`, or `camelCase`?

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala [hunk 3]
```diff
@@ -114,6 +124,60 @@ trait EffectivePolicyMutationStatements {
...
+  /** Insert actions for effective policies from resource type admin
+    */
+  private def insertEffectivePolicyActionsFromResourceTypeAdmin(newResourcePK: ResourcePK, adminTypePK: ResourceTypePK)(implicit session: DBSession): Int = {
+    val eaCol = EffectivePolicyActionTable.column
+    val pa = PolicyActionTable.syntax("pa")
+    val newResource = ResourceTable.syntax("newResource")
+    val adminResource = ResourceTable.syntax("adminResource")
+    val ra = ResourceActionTable.syntax("ra")
+    val ep = EffectiveResourcePolicyTable.syntax("ep")
+    val sourcePolicy = PolicyTable.syntax("source_policy")
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala]: Same nit.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala [hunk 4]
```diff
@@ -175,6 +239,26 @@ trait EffectivePolicyMutationStatements {
...
 
+  /** Insert effective policies from resource type admin for this resource.
+    */
+  private def insertEffectivePoliciesFromResourceTypeAdmin(newResourcePK: ResourcePK, adminTypePK: ResourceTypePK, adminResourceId: ResourceId)(implicit
+      session: DBSession
+  ): Int = {
+    val p = PolicyTable.syntax("p")
+    val epCol = EffectiveResourcePolicyTable.column
+    val adminResource = ResourceTable.syntax("adminResource")
+    val newResource = ResourceTable.syntax("newResource")
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala]: Same nit.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala [hunk 5]
```diff
@@ -175,6 +239,26 @@ trait EffectivePolicyMutationStatements {
...
+
+    samsql"""insert into ${EffectiveResourcePolicyTable.table} (${epCol.resourceId}, ${epCol.sourcePolicyId})
+        select ${newResource.id}, ${p.id}
+        from ${ResourceTable as newResource},
+        ${ResourceTable as adminResource}
+        join ${PolicyTable as p} on ${adminResource.id} = ${p.resourceId}
+        where ${adminResource.resourceTypeId} = ${adminTypePK}
+        and ${adminResource.name} = ${adminResourceId}
+        and ${newResource.id} = ${newResourcePK}""".update().apply()
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala:259]: Same nit.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala [hunk 6]
```diff
@@ -283,6 +405,40 @@ trait EffectivePolicyMutationStatements {
...
   }
 
+  protected def insertResourceTypeAdminEffectivePolicyActions(policyId: PolicyPK, resourceTypePK: ResourceTypePK)(implicit session: DBSession): Int = {
+    val eaCol = EffectivePolicyActionTable.column
+
+    val ep = EffectiveResourcePolicyTable.syntax("ep")
+    val pa = PolicyActionTable.syntax("pa")
+    val r = ResourceTable.syntax("r")
+    val ra = ResourceActionTable.syntax("ra")
+    val policy = PolicyTable.syntax("policy")
```
Comment (tlangs) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/EffectivePolicyMutationStatements.scala:415]: Same nit.

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/AccessPolicyDAO.scala (hunk 1 above)
Comment (dvoet) [reply src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/AccessPolicyDAO.scala:24]: Had to touch this code anyway to remove resource type admin effective policies from deleted resources where we want to leave a tombstone record and it bothered me too much to leave alone.

**Review by tlangs — APPROVED** [review]

**Review by Ghost-in-a-Jar — APPROVED** [review]

**Pull request merged**